### PR TITLE
Infer input size in batchnorm using aggregate axes

### LINKED ISSUE #3221: Defer initialization of size of BatchNormalization and infer it at first iteration.
The input dimension of several links (e.g. `L.Linear`, `L.Convolution`) do not have to be fed and is inferred by the minibatch size of the first iteration. Theoretically we should do the similar thing with `L.BatchNormalization`.

### PR DESCRIPTION
**Pull request by toslunar**
This PR takes over #4149.  Make `size` optional in `L.BatchNormalization` and infer it using `axis` (aggregate axes), which is added in #4266.

Close #3221.

### CODE REVIEW

**Comment by toslunar** [comment]
To the reviewer: it's ok to me to split the example document as a new PR.

File: chainer/links/normalization/batch_normalization.py [hunk 1]
```diff
@@ -68,35 +69,114 @@ class BatchNormalization(link.Link):
...
+            >>> bn.avg_mean.shape
+            (3,)
+
+            >>> bn = chainer.links.BatchNormalization(axis=0)
+            >>> y = bn(x)
+            >>> bn.avg_mean.shape
+            (3, 32, 32)
+
+            You can omit ``size`` if ``axis`` is given.
+
```
Comment (delta2323) [inline chainer/links/normalization/batch_normalization.py]: Could you write examples for the case in which `size` is `None`?

**Comment by delta2323** [comment]
Could you add unit tests to test `size=None` case.

File: chainer/links/normalization/batch_normalization.py [hunk 2]
```diff
@@ -68,35 +69,132 @@ class BatchNormalization(link.Link):
...
+            >>> bn.avg_mean.shape
+            (3,)
+            >>> y = bn(x)
+            >>> y.shape
+            (10, 3, 32, 32)
+            >>> np.testing.assert_allclose(
+            ...     y.array.mean(axis=(0, 2, 3)), 0, atol=1e-6)
+            >>> np.testing.assert_allclose(
+            ...     y.array.var(axis=(0, 2, 3)),
+            ...     x.var(axis=(0, 2, 3)) / (x.var(axis=(0, 2, 3)) + 2e-5),
```
Comment (delta2323) [inline chainer/links/normalization/batch_normalization.py]: I think `np.ones(3, dtype=np.float32)` is simpler.

Comment (delta2323) [reply chainer/links/normalization/batch_normalization.py]: I noticed that BN has `beta` and `gamma`, so, the output `y` is whitened (mean=0, var=1) only when `beta` and `gamma` are initial values or they are removed by setting  `use_gamma` and `use_beta` options to `True` (the example comes under the first case). Therefore, users might misunderstand and think `y` is always whitened if we put these assertions. (sorry, I should have noticed that earlier).

Comment (delta2323) [reply chainer/links/normalization/batch_normalization.py]: Alternative ideas are:
1. Set `use_gamma` and `use_beta` to `False`.
2. Write as `y` is whitened because `beta` and `gamma` are initial values.
In both case, I would suggest to write that `y` is not whitened in general.

**Comment by toslunar** [comment]
Fixed

**Comment by okuta** [comment]
LGTM!

**Comment by okuta** [comment]
@delta2323 Do you have any comments?

File: chainer/links/normalization/batch_normalization.py [hunk 3]
```diff
@@ -68,35 +69,139 @@ class BatchNormalization(link.Link):
...
+            By default, channel axis is (or starts from) the 1st axis of the
+            input shape.
+
+        2. Give the aggregate axes:
+
+            from Chainer v5
+
+            With ``axis`` option, similarly to NumPy, you may specify the
+            aggregate axes, which are treated as the "batch" axes for the
+            batch statistics. You can omit ``size`` if ``axis`` is given.
```
Comment (delta2323) [inline chainer/links/normalization/batch_normalization.py]: It might be helpful to write explicitly that creation of `avg_mean` and `avg_var` is deferred until first forward propagation.

**Comment by delta2323** [comment]
LGTM, waiting for the CI.

**Comment by delta2323** [comment]
jenkins, test this please.

**Comment by delta2323** [comment]
I'm working on it.

**Comment by niboshi** [comment]
When the backport of this PR is merged, the backport of #5017 should also be merged in the same release.

**Comment by delta2323** [comment]
Sorry for late I made a backport PR #5050.

**Comment by delta2323** [comment]
We decided not to backport this PR, as discussed in #5050.

**Comment by delta2323** [comment]
We decided not to backport this PR, as discussed in #5050. Please write comments or requests if you have ones.
